include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -2163,16 +2167,24 @@ namespace internal
                               dofs_per_face);
               const unsigned int *index_array =
                 &data.face_to_cell_index_nodal(face_no, 0);
+              const unsigned int *orientation =
+                &data.face_orientations[face_orientation][0];
+
               for (unsigned int i = 0; i < dofs_per_face; ++i)
-                {
-                  const unsigned int ind = index_array[i];
-                  for (unsigned int comp = 0; comp < n_components; ++comp)
-                    do_vectorized_read(src_ptr + dof_index +
-                                         (ind +
-                                          comp * static_dofs_per_component) *
-                                           VectorizedArrayType::size(),
-                                       temp1[i + 2 * comp * dofs_per_face]);
-                }
+                for (unsigned int comp = 0; comp < n_components; ++comp)
+                  do_vectorized_read(
+                    src_ptr + dof_index +
+                      (index_array[i] + comp * static_dofs_per_component) *
+                        VectorizedArrayType::size(),
+                    temp1[((dim < 3 || face_orientation == 0 ||
+                            subface_index <
+                              GeometryInfo<dim>::max_children_per_cell) ?
+                             i :
+                             orientation[i]) +
+                          2 * comp * dofs_per_face]);
```

Comment:
This is the hot path of the read operation. Have you checked that the compiler can move the checks out of the `i` loop? I would lean towards duplicating the loop because it isn't probably even that much more code to read given the difficult clause in the access.

Reply:
I agree that the if-statement is a mess. But duplicating the loop would practically lead to the duplication of the whole function. I think we need a different solution: template parameter?